plugins/modules/discovery_intent.py

```diff
@@ -465,36 +465,41 @@ def preprocessing_devices_info(self, devices_list=None):
                              of IP ranges separated by commas.
         """
 
-        if devices_list is None:
-            devices_list = []
-
-        ip_address_list = [device['ip'] for device in devices_list]
+        if ip_address_list is None:
+            ip_address_list = []
 
         self.log("Discovery type passed for the discovery is {0}".format(self.validated_config[0].get('discovery_type')), "INFO")
         if self.validated_config[0].get('discovery_type') in ["SINGLE", "CDP", "LLDP"]:
             if len(ip_address_list) == 1:
                 ip_address_list = ip_address_list[0]
             else:
-                self.log("Device list's length is longer than 1", "ERROR")
-                self.module.fail_json(msg="Device list's length is longer than 1", response=[])
+                self.log("IP Address list's length is longer than 1", "ERROR")
+                self.module.fail_json(msg="IP Address list's length is longer than 1", response=[])
         elif self.validated_config[0].get('discovery_type') == "CIDR":
             if len(ip_address_list) == 1 and self.validated_config[0].get('prefix_length'):
                 ip_address_list = ip_address_list[0]
                 ip_address_list = str(ip_address_list) + "/" + str(self.validated_config[0].get('prefix_length'))
             else:
-                self.log("Device list's length is longer than 1", "ERROR")
-                self.module.fail_json(msg="Device list's length is longer than 1", response=[])
+                self.log("IP Address list's length is longer than 1", "ERROR")
+                self.module.fail_json(msg="IP Address list's length is longer than 1", response=[])
+        elif self.validated_config[0].get('discovery_type') == "RANGE":
+            if len(ip_address_list) == 1:
+                if len(str(ip_address_list[0]).split("-")) == 2:
+                    ip_address_list = ip_address_list[0]
+                else:
+                    ip_address_list = "{0}-{1}".format(ip_address_list[0], ip_address_list[0])
+            else:
+                self.log("IP Address list's length is longer than 1", "ERROR")
+                self.module.fail_json(msg="IP Address list's length is longer than 1", response=[])
         else:
```

Comment:
As many lines are repeated in this API:

    if discovery_type in ["SINGLE", "CDP", "LLDP"]:
        if len(ip_address_list) == 1: 
            processed_ip_addresses = ip_address_list[0] 
        else:
            self.handle_discovery_error()
    elif discovery_type == "CIDR":
        processed_ip_addresses = self.process_discovery_cidr(ip_address_list)
    elif discovery_type == "RANGE":
        processed_ip_addresses = self.process_discovery_range(ip_address_list)
    else:
        processed_ip_addresses = self.process_discovery_other(ip_address_list)

    self.log("Collected IP address/addresses are {0}".format(str(processed_ip_addresses)), "INFO")

And followed by the functions?

```
def handle_discovery_error(self):
    self.log("The length of the IP address list is longer than 1", "ERROR")
    self.module.fail_json(msg="he length of the IP address list is longer than 1", response=[])

def process_cidr(self, ip_address_list):
    if  len(ip_address_list) != 1:
                self.handle_discovery_error()
                return

    prefix_length = self.validated_config[0].get('prefix_length')
#check prefix length if required...
    return "{0}/{1}".format(ip_address_list[0], prefix_length))
```

Likewise you can write other APIs...